### PR TITLE
Feature/c header and example logger

### PR DESCRIPTION
**Pull request by helixhorned**
This pull request adds a C version of the interposition macro, together with accompanying example logger / test (`make test_c`) and documentation. The macro naturally ends up being more verbose and clumsy to use.

This is useful in scenarios where it is undesirable that the generated SO has dependencies on the C++ standard library, for example because the target might have an older version of it and one does not want to risk dynamic load time incompatibilities. \
Sure, one could use `-static-libstdc++`, but in that case it may be desirable that the generated SO is as small as possible. So, one would always have to watch out whether using some feature pulls in calls to the library (i.e. is not completely inlineable). \
In short, using C is nice when one desires minimalism and C++ is unnecessary for the purposes of implementing the interposition code.

### CODE REVIEW

**Comment by helixhorned** [comment]
Edited `CLIB` -> `CCLIB` (because `CC` is the counterpart of `CXX`) in the Makefile and force-pushed.

**Comment by helixhorned** [comment]
This needs more work. Adding `-Wall -Wextra -pedantic` in followup branch `feature/more-safety` gives:

```
logger.c:28:1: warning: void function 'Real__free' should not return void expression [-Wpedantic]
INTERPOSE_C(void, free, (void* p), (p)) {
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../include/interpose.h:35:5: note: expanded from macro 'INTERPOSE_C'
    return func ARG_NAME_LIST; \
    ^      ~~~~~~~~~~~~~~~~~~
```

And right, both C99 and C11 have:

> 6.8.6.4 The `return` statement
>
> **Constraints**
>
> 1. A `return` statement with an expression shall not appear in a function whose return type is `void`. A `return` statement without an expression shall only appear in a function whose return type is `void`.

**Comment by helixhorned** [comment]
macOS version added, tested on Catalina with its default compiler (Xcode was updated just now).

**Comment by helixhorned** [comment]
And one more force-push because I got the "Based on" wrong in the first commit -- it would have been self-referential :facepalm: .

**Comment by kabicm** [comment]
This PR is great, please merge it!
